server/interfaces/interfaces.go

```diff
@@ -449,6 +451,16 @@ type MetricsCollector interface {
 	ReadCount(ctx context.Context, counterName string) (int64, error)
 }
 
+// A ProtoStore allows for storing ephemeral protos globally.
+//
+// No guarantees are made about durability of ProtoStores -- they may be
+// evicted from the backing store that maintains them (usually memcache or
+// redis), so they should *not* be used in critical path code.
+type ProtoStore interface {
+	SetMessageByKey(ctx context.Context, key string, msg proto.Message) error
+	GetMessageByKey(ctx context.Context, key string, msg proto.Message) error
+}
+
```

Comment:
I think this interface would have higher utility if it were called `KeyValueStore` with `GetString` / `SetString` methods. That way the implementations don't need to care about what's being stored. Then you could have separate utility functions like `{read,write}Proto{To,From}KeyValueStore` which handles the proto serialization

Reply:
Actually, you might be able to get away with using the existing memory cache / redis cache implementations instead of introducing a new interface -- you can serialize the proto to bytes then use `digest.ComputeForMessage` as the key

Reply:
Regarding the first point, I sort of feel like a proto already doesn't care about what's being stored. Anything, more or less, can be serialized to a proto, and protos provide much better structure than strings when it comes to serialization/deserialization. Do you disagree?

As far as the second point goes, you hit the nail on the head. I absolutely couldn't use CAS because I can't guarantee that I know the digest at retrieval time, and I actually tripped on that for a while at first before I realized I could just write an interface directly for redis/memory and implement a new cache type.

Reply:
My comment is mostly motivated by the fact that protos aren't really "native" to any of the underlying implementations (in-memory maps in golang, or Redis), while simpler datatypes like `string` and `[]byte` are. So effectively all the implementations have to include a translation layer from proto to the underlying storage mechanism, and if you want to use this to store something that isn't a proto, you have to create a "wrapper" proto to get it to work with this interface.

But more generally, the less the interface assumes about the type of data being passed in, the more powerful of an abstraction it is, since it's compatible with more things and is easier to implement. `io.Reader` and `io.Writer` for example both work with `[]byte` and they are super ubiquitous because `[]byte` is sort of the "least common denominator" among all types of data and Go makes it very convenient to work with `[]byte`. (Speaking of... the KeyValueStore interface I mentioned should probably also deal with `[]byte` for the values rather than `string`)

Reply:
I am much more onboard with using `byte[]`. It was encoding things as strings that was making me uncomfy :P

Reply:
Done!